File: aptcompiler/src/main/java/com/eugene/aptcompiler/router/RouterProcessor.java
package com.eugene.aptcompiler.router;

import com.eugene.aptannotation.router.*;
import com.eugene.aptcompiler.router.resolver.*;
import com.google.auto.service.AutoService;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import java.io.IOException;
import java.util.*;

//@AutoService(Processor.class)
public class RouterProcessor extends AbstractProcessor {

    /**
     * 文件相关的辅助类
     */
    private Filer mFiler;
    /**
     * 元素相关的辅助类
     */
    private Elements mElementUtils;
    /**
     * 日志相关的辅助类
     */
    private Messager mMessager;

    private Map<String, RouterHelper> mRouterHelpers = new HashMap<>();

    @Override
    public synchronized void init(ProcessingEnvironment processingEnvironment) {
        super.init(processingEnvironment);
        mFiler = processingEnvironment.getFiler();
        mElementUtils = processingEnvironment.getElementUtils();
        mMessager = processingEnvironment.getMessager();
//        mMessager.printMessage(Diagnostic.Kind.ERROR,"日志开始---------------");
    }

    /**
     * @Return 指定哪些注解应该被注解处理器注册
     */
    @Override
    public Set<String> getSupportedAnnotationTypes() {
        Set<String> types = new LinkedHashSet<>();
        types.add(ToThis.class.getCanonicalName());
        types.add(Extra.class.getCanonicalName());
        return types;
    }

    /**
     * @return 指定使用的 Java 版本。通常返回 SourceVersion.latestSupported()。
     */
    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }

    @Override
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnvironment) {
        mRouterHelpers.clear();

        processFieldExtra(roundEnvironment);
        try {
            for (RouterHelper helper : mRouterHelpers.values()) {
                helper.generateResolve().writeTo(mFiler);
            }
        } catch (IOException e) {
            e.printStackTrace();
            return true;
        }

        return true;
    }

    private void processFieldExtra(RoundEnvironment roundEnv) {

        for (Element element : roundEnv.getElementsAnnotatedWith(ToThis.class)) {
            RouterHelper helper = createRouterHelper(element);
            ToThisResolver codeResolver = new ToThisResolver(element);
            helper.setToThisResolver(codeResolver);
        }

        for (Element element : roundEnv.getElementsAnnotatedWith(Extra.class)) {
            RouterHelper helper = getRouterHelperByExtra(element);
            ExtraResolver extra = new ExtraResolver(element);
            helper.addField(extra);
        }

    }

    private RouterHelper createRouterHelper(Element element) {
        TypeElement classElement = (TypeElement) element;
        return getRouterHelper(classElement);
    }

    private RouterHelper getRouterHelperByExtra(Element element) {
        TypeElement classElement = (TypeElement) element.getEnclosingElement();
        return getRouterHelper(classElement);
    }

    private RouterHelper getRouterHelper(TypeElement classElement) {
        String fullClassName = classElement.getQualifiedName().toString();
        RouterHelper helper = mRouterHelpers.get(fullClassName);
        if (helper == null) {
            helper = new RouterHelper(mElementUtils, classElement);
            mRouterHelpers.put(fullClassName, helper);
        }
        return helper;
    }

}

File: aptcompiler/src/main/java/com/eugene/aptcompiler/krouter/RouterProcessor.kt
package com.eugene.aptcompiler.krouter

import com.eugene.aptannotation.router.Extra
import com.eugene.aptannotation.router.ToThis
import com.eugene.aptcompiler.krouter.resolver.ExtraResolver
import com.eugene.aptcompiler.krouter.resolver.ToThisResolver
import com.squareup.kotlinpoet.FileSpec
import java.io.File
import java.io.IOException
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import javax.tools.Diagnostic

class RouterProcessor : AbstractProcessor() {

    /**
     * 文件相关的辅助类
     */
    private var mFiler: Filer? = null
    /**
     * 元素相关的辅助类
     */
    private var mElementUtils: Elements? = null
    /**
     * 日志相关的辅助类
     */
    private var mMessager: Messager? = null

    private val mRouterHelpers = HashMap<String, RouterHelper>()

    override fun init(processingEnvironment: ProcessingEnvironment?) {
        super.init(processingEnvironment)
        mFiler = processingEnvironment?.filer
        mElementUtils = processingEnvironment?.elementUtils
        mMessager = processingEnvironment?.messager
        mMessager?.printMessage(Diagnostic.Kind.WARNING, "日志开始---------------")
    }

    override fun getSupportedAnnotationTypes(): MutableSet<String> {
        mMessager?.printMessage(Diagnostic.Kind.WARNING, "getSupportedAnnotationTypes---------------")
        return mutableSetOf<String>().apply {
            this.add(ToThis::class.java.canonicalName)
            this.add(Extra::class.java.canonicalName)
        }
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.latestSupported()
    }

    override fun process(set: MutableSet<out TypeElement>?, roundEnvironment: RoundEnvironment?): Boolean {
        mRouterHelpers.clear()
        mMessager?.printMessage(Diagnostic.Kind.WARNING, "process---------------")
        processFieldExtra(roundEnvironment)

        for (helper in mRouterHelpers.values) {
            helper.generateResolve().writeTo(mFiler!!)
        }

        return true
    }

    private fun processFieldExtra(roundEnv: RoundEnvironment?) {

        roundEnv?.getElementsAnnotatedWith(ToThis::class.java)
            ?.forEach {
                val helper = createRouterHelper(it)
                val codeResolver = ToThisResolver(it)
                helper.setToThisResolver(codeResolver)
            }

        roundEnv?.getElementsAnnotatedWith(Extra::class.java)
            ?.forEach {
                val helper = getRouterHelperByExtra(it)
                val extra = ExtraResolver(it)
                helper.addField(extra)
            }
    }

    private fun createRouterHelper(element: Element): RouterHelper {
        val classElement = element as TypeElement
        return getRouterHelper(classElement)
    }

    private fun getRouterHelperByExtra(element: Element): RouterHelper {
        val classElement = element.enclosingElement as TypeElement
        return getRouterHelper(classElement)
    }

    private fun getRouterHelper(classElement: TypeElement): RouterHelper {
        val fullClassName = classElement.qualifiedName.toString()
        var helper = mRouterHelpers[fullClassName]
        if (helper == null) {
            helper = RouterHelper(mElementUtils!!, classElement)
            mRouterHelpers[fullClassName] = helper
        }
        return helper
    }


}
